src/test/java/uk/gov/pay/adminusers/persistence/dao/GovUkPayAgreementDaoTest.java

```diff
@@ -0,0 +1,88 @@
+
+import org.junit.Before;
+import org.junit.Test;
+import uk.gov.pay.adminusers.model.Service;
+import uk.gov.pay.adminusers.persistence.entity.GovUkPayAgreementEntity;
+
+import javax.persistence.RollbackException;
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static uk.gov.pay.adminusers.app.util.RandomIdGenerator.randomInt;
+import static uk.gov.pay.adminusers.fixtures.ServiceDbFixture.serviceDbFixture;
+import static uk.gov.pay.adminusers.fixtures.GovUkPayAgreementDbFixture.govUkPayAgreementDbFixture;
+
+public class GovUkPayAgreementDaoTest extends DaoTestBase{
+    
+    private GovUkPayAgreementDao agreementDao;
+    
+    @Before
+    public void setUp() {
+        agreementDao = env.getInstance(GovUkPayAgreementDao.class);
+    }
+    
+    @Test
+    public void shouldPersistEntity() {
+        int serviceId = serviceDbFixture(databaseHelper)
+                .withGatewayAccountIds(randomInt().toString())
+                .insertService()
+                .getId();
+
+        GovUkPayAgreementEntity entity = new GovUkPayAgreementEntity(serviceId, 
+                "someone@example.org", 
+                LocalDateTime.now());
+        
+        agreementDao.persist(entity);
+        assertThat(entity.getId(), is(notNullValue()));
+        List<Map<String, Object>> searchResults = databaseHelper.findGovUkPayAgreementEntity(entity.getId());
+        assertThat(searchResults.size(), is(1));
+        assertThat(searchResults.get(0).get("service_id"), is(serviceId));
+        assertThat(searchResults.get(0).get("email"), is("someone@example.org"));
+        assertThat(((Timestamp) searchResults.get(0).get("agreement_time")).toLocalDateTime(), is(entity.getAgreementTime()));
+    }
+
+    @Test
+    public void shouldNotPersistStripeAgreementEntityWhenOneAlreadyExistsForService() {
+        int serviceId = serviceDbFixture(databaseHelper)
+                .withGatewayAccountIds(randomInt().toString())
+                .insertService()
+                .getId();
+
+        GovUkPayAgreementEntity entity = new GovUkPayAgreementEntity(serviceId,
+                "someone@example.org",
+                LocalDateTime.now());
+        agreementDao.persist(entity);
+
+        GovUkPayAgreementEntity anotherEntity = new GovUkPayAgreementEntity(serviceId,
+                "someone.else@example.org",
+                LocalDateTime.now());
+        
+        try {
+            agreementDao.persist(anotherEntity);
+            fail();
+        } catch (RollbackException e) {
+            assertTrue(e.getMessage().contains("Key (service_id)=(" + serviceId + ") already exists."));
```

Comment:
Would it be better to use the ExpectedException rule to assert that the exception is thrown in this case?
If the exception wasn't thrown it wouldn't reach this line so the test would still pass.

Reply:
We can